### ISSUE TITLE
Need better understanding about how the whole thing works

### ISSUE THREAD

**Issue by green-anger** [Stability-AI/StableSwarmUI#144]
Right now it's a big mess in my head about how StableSwarm is supposed to and does currently work. I've a few questions, they're kind of related, so I don't want to create a separate issue for each. In no particular order:

On the title page (readme) it says StableSwarm is UI. In the [motivations doc](https://github.com/Stability-AI/StableSwarmUI/blob/master/docs/Motivations.md) it says about choosing C# as the backend server. Backend for UI (considering StableSwarm doesn't generate anything by itself)? 

In the same motivations doc it says about "main" backend and using Comfy for that. Two backends? Or your front talking to your backend which in turn talking to Comfy?

Then it says about choosing C# "...to maximize performance while only minimally increasing code complexity. While most ML projects tend to be written in Python..." - Comfy is written in Python + JS which you use as the "main" backend. So, you cannot maximize performance in this case if you let Comfy do all the work. What role does C# code play in here?

What IS Comfy backend? Their repo has no tags or releases or mature docs, but there must be some API to use it as backend. Does it mean you're limited to what Comfy gives access to in there and cannot extend it in no way? What benefits of using StableSwarm over vanilla Comfy, better UI?

From other issues and part "Swarm" in the name I assume it can utilize several GPUs/machines. Is it where C# comes into play? What about when used in a single machine with a single GPU, is this "swarm" part just ignored and there's no gain in the performance department?

Hope it's not too much. Thanks!

**Comment by mcmonkey4eva** [comment]
`On the title page (readme) it says StableSwarm is UI. In the [motivations doc](https://github.com/Stability-AI/StableSwarmUI/blob/master/docs/Motivations.md) it says about choosing C# as the backend server. Backend for UI (considering StableSwarm doesn't generate anything by itself)?` phrasing mixup, my bad - "backend server" in C# would rather be the mid-layer server, ie the thing between the frontend (the website) and the backend (comfy), that manages orchestration of requests, secondary processing task, file/model/etc. handling, etc.

This and the other confusions with that doc are made weird by the fact that "comfy as the backend" wasn't a locked in thing at the time it was written, the original goal was to implement the backend directly in C# but leave it modular. It ended up better to focus in on Comfy despite the slight performance disadvantage as it benefits too many other areas to give up easily, at least for now.

`Does it mean you're limited to what Comfy gives access to in there and cannot extend it in no way?` Have you seen, uh, what comfy is? It's a nodegraph processor where nodes are easily externally defined. Swarm has a variety of custom nodes it uses and of course generates the graphs themselves on the fly (in C#).

`What benefits of using StableSwarm over vanilla Comfy, better UI?` Yes. If you use vanilla Comfy you just get a view of the nodegraph and a thumbs up... if you use Swarm, you get a cleaner interface, advanced tools, etc. (and the raw comfy graphs can be accessed and mucked with if you really want to still, but you don't *have to*)

`From other issues and part "Swarm" in the name I assume it can utilize several GPUs/machines.` Yes

`Is it where C# comes into play?` This is one of the many things being handled in the C# code yes. The C# code is the largest part of the project and does a ton.

`What about when used in a single machine with a single GPU, is this "swarm" part just ignored and there's no gain in the performance department?` there's still performance advantages even on a single GPU, especially when compared to easy to use UI options that aren't based on comfy (as comfy itself has performance advantages over the other options as well)

**Comment by green-anger** [comment]
`Have you seen, uh, what comfy is? It's a nodegraph processor where nodes are easily externally defined. Swarm has a variety of custom nodes it uses and of course generates the graphs themselves on the fly (in C#).`

Yes, I've been using Comfy for a while including custom nodes and building custom workflows with adetailer, upscaling, controlnet and such. I've never dived into the Comfy code, but as I understand it uses [litegraph.js](https://github.com/jagenjo/litegraph.js/) as its front which converts all the info into json (likely) which is pushed to Python for parsing and all manipulations. You can, of course, add your own custom nodes but it must be compliant with the system which has its limitation (as any system), e.g. horrible split of numbers into int/float/number which are considered to be different and not compatible. And when you say `Swarm has a variety of custom nodes` do you mean Comfy custom node or is it another set of nodes that must be written separately from Comfy nodes? Does it bypass Comfy conversion process and can be pushed to Comfy backend directly? Are they compatible with Comfy nodes or is it a completely different set?

`there's still performance advantages even on a single GPU, especially when compared to easy to use UI options that aren't based on comfy`

But how? Comfy UI is clunky at best, I'd even dare to say it's almost non-existent, as you very correctly noted it's a `nodegraph processor`, so all you can do is connect nodes to each other. Even looking at the screenshots I can say Swarm does much better job on that part and if you meant advantages in that, then it's good. But once I've build my flow in Comfy I can change the prompt, few parameters and generate, the same as in Swarm or a1111. And since you add a middle layer it cannot improve speed, only slow it down by doing some extras, because you use the same Comfy backend that accepts data in certain format you must comply with.

`The C# code is the largest part of the project and does a ton.`

That's why I asked before what Comfy backend was. If the middle layer does a lot of work, the result of that work must be sent to Comfy backend. And that's where you must have some API and where the system limitation comes from. Or did you just rip out some code from Comfy and use it in Swarm? Does it mean you can extend the core backend and improve it? For example, I don't know if it's possible in the Comfy backend (vanilla UI doesn't give such option or at least I was so blind that didn't see it), but it'd be great to be able to load several models in RAM and see which are loaded, that'd improve speed of switching between them. The last version of a1111 I used allowed loading several models but didn't show which were loaded, I assume it took the latest you selected. It should be possible to add several checkpoint loaders in Comfy UI and if they're all used in the same flow they'll probably all be loaded, but I'm interested in the use case with one loader and switching between models with no delay in generation after switching. Is it (or will it be) possible in Swarm?

**Comment by mcmonkey4eva** [comment]
`e.g. horrible split of numbers into int/float/number which are considered to be different and not compatible.` that's broadly a limitation of convenience/ease-of-use for those interacting with the litegraph UI, and not something generally in the way of software based usages like what Swarm does. Number type things in comfy are irrelevant to Swarm as Swarm does its own basic number processing and doesn't need the node graph to do that.

`do you mean Comfy custom node or is it another set of nodes that must be written separately from Comfy nodes? Does it bypass Comfy conversion process and can be pushed to Comfy backend directly? Are they compatible with Comfy nodes or is it a completely different set?` Custom comfy nodes, in the same was as you've used custom comfy nodes for adetailer, controlnet, etc. -- https://github.com/Stability-AI/StableSwarmUI/tree/master/src/BuiltinExtensions/ComfyUIBackend/ExtraNodes 

You mention adetailer, and that's a good example here - Swarm has a more advanced adetailer-like system built in:
You can, in the prompt box of the generate tab, simply type `a photo of a person   <segment:face> an hd face` and C# code parses that prompt input, and reformats a workflow that contains prompt `a photo of a person`,  the swarm custom clipseg node <https://github.com/Stability-AI/StableSwarmUI/blob/master/src/BuiltinExtensions/ComfyUIBackend/ExtraNodes/SwarmClipSeg.py> with `face` as its input, several mask postprocessing steps, and then a swarmksampler with `an hd face` input to regenerate the masked region, and a few more postprocessing nodes to stick the result back together. This is designed to be intercompatible with anything else you can do in swarm's generate tab inputs, including other prompt tricks like `<region:>` or `<object:>`
The frontend UI (javascript) just took some text and passed it forward. The C# did the complex step of taking that input and figuring out how to process your request best and build a plan of action. Comfy's node processor took the nodes swarm gave and scheduled linear execution. Swarm's custom comfy nodes in python then did most of the specific processing, and then passed along to comfy's optimized diffusion execution internals to actually run the model as best as possible.
I often use the word "orchestration" in regards to how Swarm works, because it is a complex dance of many different components being organized - without any one of these components the job would be much harder.

`And since you add a middle layer it cannot improve speed, only slow it down by doing some extras, because you use the same Comfy backend that accepts data in certain format you must comply with.` not actually true - swarm isn't a useless middleman just sitting in between passing jobs forward, it does actual work, and not just power features like the one explained above.
For example, when you use comfy directly, at the end of the graph, you get to a "Save Image" node and the whole comfy graph waits half a second for your data drive to get the image data, save it, then to read that data back from drive and display it to the UI. (PreviewImage does this too, due to weird technical requirements in comfy for the litegraph UI - which for the record don't apply to swarm).
Meanwhile in swarm, SaveImage is automatically replaced to `SwarmSaveImageWS` which ignores the datadrive and just sends the raw output straight to swarm. Swarm's multithreaded queue handling immediately starts comfy's next generation job, while separately the image is stored on drive and shown to the UI view.
Let's say hypothetically you have a quick 2 second image gen, and a slow 0.5 second image save+load time, and you're generating 10 images. In Comfy you'll have (2 + 0.5) * 10 = 25 seconds. In Swarm you'll have 2 * 10 = 20 second gen because that 0.5 is in parallel. (Those numbers are pulled from nowhere, to explain the concept, actual times in practice depends on hardware and all that).

That's just one example, there's many little tricks Swarm can pull to accelerate things beyond even comfy's speed. But, again, it's much more impressive when you compare apples-to-apples ie compare to other user-friendly UIs like auto webui, where swarm's speed advantage is more significant *because* comfy is a core component.

`it'd be great to be able to load several models in RAM and see which are loaded` this is one of the things I wanted to have with the C# backend that I let sit as a problem for now, the advantages of comfy outweigh the limitations against tricks like that.
Comfy *does* allow for caching multiple models in memory - I'm not sure offhand how precisely to do it, but I know I've heard model caching discussed in relation to comfy before.
In Swarm you can at least add multiple backends to do that rn - that's a rather janky way to do it, but it works, mostly.
If you'd like it added as a more dedicated feature, please post an issue specifically as a feature request for that and it can probably be built.
I've thought a bit about smarter multi-model select/loading UI control but haven't spent too much time on it thus far.

**Comment by green-anger** [comment]
I appreciate your thorough replies, it was a very insightful conversation! 🤝 

Opened a new [feature request](https://github.com/Stability-AI/StableSwarmUI/issues/146) as discussed.
